File: plugins/OrchidWiki/src/main/java/com/eden/orchid/wiki/WikiGenerator.java
package com.eden.orchid.wiki;

import com.caseyjbrooks.clog.Clog;
import com.eden.common.util.EdenUtils;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.generators.OrchidGenerator;
import com.eden.orchid.api.options.annotations.Description;
import com.eden.orchid.api.options.annotations.Option;
import com.eden.orchid.api.options.OptionsHolder;
import com.eden.orchid.api.options.annotations.StringDefault;
import com.eden.orchid.api.resources.resource.OrchidResource;
import com.eden.orchid.api.resources.resource.StringResource;
import com.eden.orchid.api.theme.pages.OrchidPage;
import com.eden.orchid.utilities.OrchidUtils;
import com.eden.orchid.wiki.model.WikiModel;
import com.eden.orchid.wiki.model.WikiSection;
import com.eden.orchid.wiki.pages.WikiPage;
import com.eden.orchid.wiki.pages.WikiSummaryPage;
import org.apache.commons.io.FilenameUtils;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.File;
import java.util.ArrayList;
import java.util.List;

@Singleton
@Description("Create a structured and navigable knowledge-base for your project.")
public class WikiGenerator extends OrchidGenerator implements OptionsHolder {

    private final WikiModel model;

    @Option("baseDir")
    @StringDefault("wiki")
    public String wikiBaseDir;

    @Option("sections")
    public String[] sectionNames;

    @Inject
    public WikiGenerator(OrchidContext context, WikiModel model) {
        super(context, "wiki", 700);
        this.model = model;
    }

    @Override
    public List<? extends OrchidPage> startIndexing() {
        model.initialize();

        if (EdenUtils.isEmpty(sectionNames)) {
            WikiSection wiki = getWikiPages(null);
            if (wiki != null) {
                model.getSections().put(null, wiki);
            }
        }
        else {
            for (String section : sectionNames) {
                WikiSection wiki = getWikiPages(section);
                if (wiki != null) {
                    model.getSections().put(section, wiki);
                }
            }
        }

        return model.getAllPages();
    }

    @Override
    public void startGeneration(List<? extends OrchidPage> pages) {
        pages.forEach(context::renderTemplate);
    }

    private WikiSection getWikiPages(String section) {
        JSONObject pageMenuItem = new JSONObject();
        pageMenuItem.put("type", "wiki");

        if(!EdenUtils.isEmpty(section)) {
            pageMenuItem.put("title", section + " Wiki");
            pageMenuItem.put("section", section);
        }
        else {
            pageMenuItem.put("title", "Wiki");
        }

        ArrayList<WikiPage> wiki = new ArrayList<>();

        String sectionBaseDir = (!EdenUtils.isEmpty(section)) ?
                OrchidUtils.normalizePath(wikiBaseDir) + "/" + OrchidUtils.normalizePath(section) + "/" :
                OrchidUtils.normalizePath(wikiBaseDir) + "/";

        OrchidResource summary = context.getLocalResourceEntry(sectionBaseDir + "SUMMARY.md");

        if (summary == null) {
            Clog.w("Could not find wiki summary page in '#{}'", sectionBaseDir);
            return null;
        }

        String content = context.compile(summary.getReference().getExtension(), summary.getContent());
        Document doc = Jsoup.parse(content);

        Elements links = doc.select("a[href]");

        WikiPage previous = null;

        int i = 0;

        for (Element a : links) {
            String file = sectionBaseDir + a.attr("href");
            String path = sectionBaseDir + FilenameUtils.removeExtension(a.attr("href"));

            OrchidResource resource = context.getLocalResourceEntry(file);

            if (resource == null) {
                Clog.w("Could not find wiki resource page at '#{$1}'", file);
                resource = new StringResource(context, path + File.separator + "index.md", a.text());
            }

            WikiPage page = new WikiPage(resource, a.text());

            page.getMenu().addMenuItem(pageMenuItem);

            i++;
            page.setOrder(i);

            wiki.add(page);

            if (previous != null) {
                previous.setNext(page);
                page.setPrevious(previous);

                previous = page;
            }
            else {
                previous = page;
            }

            a.attr("href", page.getReference().toString());
        }

        String safe = doc.toString();
        summary = new StringResource(context, sectionBaseDir + "summary.md", safe);

        String sectionTitle = (!EdenUtils.isEmpty(section)) ? section : "Wiki";
        WikiSummaryPage summaryPage = new WikiSummaryPage(summary, OrchidUtils.camelcaseToTitleCase(sectionTitle));
        summaryPage.getReference().setUsePrettyUrl(true);
        summaryPage.getMenu().addMenuItem(pageMenuItem);

        for(WikiPage wikiPage : wiki) {
            wikiPage.setSectionSummary(summaryPage);
        }

        return new WikiSection(section, summaryPage, wiki);
    }

}

File: plugins/OrchidWiki/src/main/kotlin/com/eden/orchid/wiki/WikiGenerator.kt
package com.eden.orchid.wiki

import com.eden.common.util.EdenUtils
import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.generators.FileCollection
import com.eden.orchid.api.generators.OrchidCollection
import com.eden.orchid.api.generators.OrchidGenerator
import com.eden.orchid.api.generators.PageCollection
import com.eden.orchid.api.options.annotations.Description
import com.eden.orchid.api.options.annotations.ImpliedKey
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.resources.resource.StringResource
import com.eden.orchid.api.theme.pages.OrchidPage
import com.eden.orchid.wiki.model.WikiModel
import com.eden.orchid.wiki.model.WikiSection
import com.eden.orchid.wiki.pages.WikiBookPage
import com.eden.orchid.wiki.pages.WikiSectionsPage
import com.eden.orchid.wiki.utils.WikiUtils
import javax.inject.Inject

@Description("Create a structured and navigable knowledge-base for your project.", name = "Wiki")
class WikiGenerator : OrchidGenerator<WikiModel>(GENERATOR_KEY, PRIORITY_EARLY) {

    companion object {
        const val GENERATOR_KEY = "wiki"
    }

    @Option
    @ImpliedKey("key")
    @Description("The sections within the baseDir to make wikis out of.")
    lateinit var sections: MutableList<WikiSection>

    @Option
    @Description("The configuration for the default wiki, when no other categories are set up.")
    lateinit var defaultConfig: WikiSection

    override fun startIndexing(context: OrchidContext): WikiModel {
        if (EdenUtils.isEmpty(sections)) {
            sections.add(defaultConfig)
        }

        val loadedSections = ArrayList<WikiSection>()
        this.sections.forEach { section ->
            val loadedSectionContent = section.adapter.loadWikiPages(section)
            if (loadedSectionContent != null) {
                val (summaryPage, wiki) = loadedSectionContent

                WikiUtils.linkWikiPages(summaryPage, wiki)

                section.summaryPage = summaryPage
                section.wikiPages = wiki
                section.bookPage = WikiUtils.createWikiPdf(section)

                loadedSections.add(section)
            }
        }

        val model = WikiModel(loadedSections)

        if (loadedSections.size > 1) {
            model.sectionsPage = getSectionsIndex(context, model)
        }

        return model
    }

    override fun startGeneration(context: OrchidContext, model: WikiModel) {
        model.allPages.forEach {
            if (it is WikiBookPage) {
                context.renderBinary(it)
            } else {
                context.renderTemplate(it)
            }
        }
    }

    private fun getSectionsIndex(context: OrchidContext, model: WikiModel): WikiSectionsPage {
        val resource = StringResource(context, "wiki.md", "")

        val sectionsPage = WikiSectionsPage(model, resource, "Wiki")

        for (summaryPage in model.sections.values) {
            summaryPage.summaryPage.sectionsPage = sectionsPage
            summaryPage.summaryPage.parent = sectionsPage
        }

        return sectionsPage
    }

    override fun getCollections(context: OrchidContext, model: WikiModel): List<OrchidCollection<*>> {
        val collectionsList = java.util.ArrayList<OrchidCollection<*>>()

        model.sections.forEach {
            val sectionPages = ArrayList<OrchidPage>()

            sectionPages.add(it.value.summaryPage)
            sectionPages.addAll(it.value.wikiPages)

            val collection = FileCollection(this, it.key, sectionPages)
            collectionsList.add(collection)
        }

        val bookPages = model.sections.values.mapNotNull { it.bookPage }
        collectionsList.add(PageCollection(this, "books", bookPages))

        return collectionsList
    }

}

